### PR TITLE
dollarmath: Add `allow_blank_lines` option

### LINKED ISSUE #45: Do not allow blank lines within $$
See https://github.com/executablebooks/markdown-it-dollarmath/pull/8

### PR DESCRIPTION
**Pull request by eric-wieser**
Closes #45 behind a non-default `allow_blank_lines=False` flag.

Note this means that the default behavior of the javascript and python implementations has diverged.

### CODE REVIEW

**Comment by eric-wieser** [comment]
I think a maintainer (@chrisjsewell?) has to click some sort of approval button in order for CI to run, to verify that I'm not mining cryptocurrency in this PR?

File: mdit_py_plugins/dollarmath/index.py
Comment (chrisjsewell) [inline mdit_py_plugins/dollarmath/index.py:305]: I think this should stay (but break not continue) to safeguard against any index errors

**Review by chrisjsewell — CHANGES_REQUESTED** [review]

File: mdit_py_plugins/dollarmath/index.py
Comment (eric-wieser) [reply mdit_py_plugins/dollarmath/index.py:305]: This line doesn't exist in the javascript version, why would you want to have it in the Python version?

**Comment by eric-wieser** [comment]
@chrisjsewell: this is what I am referring to:

image

File: tests/fixtures/dollar_math.md
Comment (eric-wieser) [reply tests/fixtures/dollar_math.md:552]: https://github.com/executablebooks/markdown-it-dollarmath/pull/9

File: mdit_py_plugins/dollarmath/index.py
Comment (chrisjsewell) [reply mdit_py_plugins/dollarmath/index.py:305]: Because JavaScript does not act the same as Python; if you index a JavaScript array with "bad indexes" it will simply return null, whereas in python it will raise an indexerror.
There has been multiple bug fixed for this in markdown-it-py

File: tests/fixtures/dollar_math.md
Comment (chrisjsewell) [reply tests/fixtures/dollar_math.md:552]: This is a separate issue and does not need to be removed here

File: mdit_py_plugins/dollarmath/index.py
Comment (eric-wieser) [reply mdit_py_plugins/dollarmath/index.py:305]: The only indexing that takes place is `state.src[start:end]`, and that will never give an `IndexError`

**Comment by eric-wieser** [comment]
I assume the CI failure due to a bad interaction between click and black is bogus? Thanks for triggering the CI run.

File: tests/fixtures/dollar_math.md
Comment (eric-wieser) [reply tests/fixtures/dollar_math.md:552]: Sure, I'll revert this change. Can you approve CI on the PR linked above so that we can see if this test passes in the JavaScript version?

Comment (eric-wieser) [reply tests/fixtures/dollar_math.md]: Done!

File: mdit_py_plugins/dollarmath/index.py
Comment (eric-wieser) [reply mdit_py_plugins/dollarmath/index.py:305]: @rowanc1: what can we do to progress here?

**Comment by eric-wieser** [comment]
Can I do anything to help this to go through?

**Comment by chrisjsewell** [comment]
> Can I do anything to help this to go through?

I feel this behaviour needs to be under a options flag, (a) because it is a breaking change and (b) because

> This matches the behavior of LaTeX (and of the stackexchange and github markdown parsers), and prevents runaway parsing if a $$ is opened but not closed.

I would suggest this is not always the desired behaviour.
Firstly because in MyST they are intended to be a shorthand for: https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-math, which can have blank lines (and quite probably there is already people using this feature),
and (b) "runaway parsing" is actually a feature of Markdown not a bug, both fence blocks https://spec.commonmark.org/0.30/#example-126, and (extended) markdown divs https://htmlpreview.github.io/?https://github.com/jgm/djot/blob/master/doc/syntax.html#div are both "runaway" (and there is design reasons for this)

**Comment by chrisjsewell** [comment]
I think there is also a potentially critical difference between block level `$$` and inline `$$`;
 inline `$$` already does not allow blank lines

**Comment by eric-wieser** [comment]
Does this mean you would advocate for the change to be reverted on the javascript side?

**Comment by chrisjsewell** [comment]
> Does this mean you would advocate for the change to be reverted on the javascript side?

Oh I'm not touching lol. Just let's add it in here as a feature flag, and I think that's a good compromise

**Comment by eric-wieser** [comment]
> I would suggest this is not always the desired behaviour in MyST Firstly because in MyST they are partially intended to be a shorthand for: https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-math, which can have blank lines (and quite probably there is already people using this feature)

The semantics of `.. math::` is to [auto-insert an `align` environment](https://github.com/sphinx-doc/sphinx/blob/ba080286b06cb9e0cadec59a6cf1f96aa11aef5a/sphinx/util/math.py#L35-L55) when multiple lines are provided. This is pretty non-standard for a latex parser, possibly suggesting that the MyST behavior ought to be behind a flag

> and (b) "runaway parsing" is actually a feature of Markdown not a bug, both fence blocks https://spec.commonmark.org/0.30/#example-126, and (extended) markdown div blocks https://htmlpreview.github.io/?https://github.com/jgm/djot/blob/master/doc/syntax.html#div are both "runaway" (and there are design reasons for this)

I agree; but on the other hand, it seems more valuable to match what the consensus seems to be among major implementations, rather than extrapolate from a spec that doesn't actually specify the behavior here!


Can I convince you of a feature flag where the default is the new behavior, or is that going to break too many things?

**Comment by chrisjsewell** [comment]
> Can I convince you of a feature flag where the default is the new behavior, or is that going to break too many things?

Yeh just to much possibility for breakage for the outset. Of course once it is in there then it's easy to control, if defaults need to be changed

**Comment by eric-wieser** [comment]
Alright, I've added the flag to default to the old behavior.

**Comment by eric-wieser** [comment]
Note that the tests won't run until you approve them:

image

**Review by chrisjsewell — APPROVED** [review]
Cheers Eric

**Pull request merged**
